snapcraft/tests/plugins/test_nodejs.py

```diff
@@ -291,6 +298,76 @@ class Options:
         self.assertFalse(os.path.exists(plugin._npm_dir))
 
 
+class NodePluginManifestTestCase(NodePluginBaseTestCase):
+
+    scenarios = multiply_scenarios(
+        [
+            ('simple', dict(return_value=(
+                '{"dependencies": {'
+                '   "testpackage1": {"version": "1.0"},'
+                '   "testpackage2": {"version": "1.2"}}}'))),
+            ('nested', dict(return_value=(
+                '{"dependencies": {'
+                '   "testpackage1": {'
+                '      "version": "1.0",'
+                '      "dependencies": {'
+                '        "testpackage2": {"version": "1.2"}}}}}'))),
+            ('missing', dict(return_value=(
+                '{"dependencies": {'
+                '   "testpackage1": {"version": "1.0"},'
+                '   "testpackage2": {"version": "1.2"},'
+                '   "missing": {"noversion": "dummy"}}}')))],
+        [('npm', dict(package_manager='npm')),
+         ('yarn', dict(package_manager='yarn'))])
+
+    def test_get_manifest_with_node_packages(self):
+        self.run_output_mock.return_value = self.return_value
+
+        class Options:
+            source = '.'
+            node_packages = []
+            node_engine = '4'
+            npm_run = []
+            node_package_manager = self.package_manager
+        plugin = nodejs.NodePlugin('test-part', Options(),
+                                   self.project_options)
+        os.makedirs(plugin.sourcedir)
+
+        plugin.build()
+
+        self.assertThat(
+            plugin.get_manifest(), Equals(
+                collections.OrderedDict(
+                    {'node-packages':
+                     ['testpackage1=1.0', 'testpackage2=1.2']})))
+
+
+class NodePluginYarnLockManifestTestCase(NodePluginBaseTestCase):
+
+    def test_get_manifest_with_yarn_lock_file(self):
+        class Options:
+            source = '.'
+            node_packages = []
+            node_engine = '4'
+            npm_run = []
+            node_package_manager = 'yarn'
+        plugin = nodejs.NodePlugin('test-part', Options(),
+                                   self.project_options)
+        os.makedirs(plugin.sourcedir)
+        with open(os.path.join(
+                plugin.sourcedir,
+                'yarn.lock'), 'w') as yarn_lock_file:
+            yarn_lock_file.write('test yarn lock contents')
+
+        plugin.build()
+
+        self.assertThat(
+            plugin.get_manifest(), Equals(
+                collections.OrderedDict(
+                    {'yarn-lock-contents': 'test yarn lock contents',
```

Comment:
The fact that the test can get away with arbitrary data makes me think the plugin should at least do basic validation - but then it took all of 5 minutes in DuckDuckGo to find that it's not YAML but some 500 lines of custom parsing...

So, I think some validation would be nice, whether that's a regex or nodejs command, before saving the lock contents in the manifest. And using a real-looking example in the test. But I'll leave this as a would-be-nice suggestion.

Reply:
If the plugin can build the package with the contents of the yarn.lock, we should record whatever was used despite of the specific value. If the value is not valid, the plugin should fail before recording anything.